src/ott/solvers/nn/neuraldual.py

```diff
@@ -108,55 +146,138 @@ def setup(
+          train=True, to_optimize="both"
+      )
+      self.valid_step_parallel = self.get_step_fn(
+          train=False, to_optimize="both"
+      )
+      self.train_fn = self.train_neuraldual_parallel
+    else:
+      if self.parallel_updates:
+        warnings.warn(
+            'parallel_updates set to True but disabling it because num_inner_iters>1'
+        )
+      self.train_step_f = self.get_step_fn(train=True, to_optimize="f")
+      self.valid_step_f = self.get_step_fn(train=False, to_optimize="f")
+      self.train_step_g = self.get_step_fn(train=True, to_optimize="g")
+      self.valid_step_g = self.get_step_fn(train=False, to_optimize="g")
+      self.train_fn = self.train_neuraldual_alternating
 
   def __call__(
       self,
       trainloader_source: Iterable[jnp.ndarray],
       trainloader_target: Iterable[jnp.ndarray],
       validloader_source: Iterable[jnp.ndarray],
       validloader_target: Iterable[jnp.ndarray],
-  ) -> Union[potentials.DualPotentials, Tuple[potentials.DualPotentials,
-                                              Train_t]]:
-    logs = self.train_neuraldual(
+      callback: Optional[Callable] = None,
+  ) -> Union[Potentials_t, Tuple[Potentials_t, Train_t]]:
+    logs = self.train_fn(
         trainloader_source,
         trainloader_target,
         validloader_source,
         validloader_target,
+        callback=callback,
     )
     res = self.to_dual_potentials()
 
     return (res, logs) if self.logging else res
 
-  def train_neuraldual(
+  def train_neuraldual_parallel(
+      self,
+      trainloader_source: Iterable[jnp.ndarray],
+      trainloader_target: Iterable[jnp.ndarray],
+      validloader_source: Iterable[jnp.ndarray],
+      validloader_target: Iterable[jnp.ndarray],
+      callback: Optional[Callable] = None,
+  ) -> Train_t:
+    """Implementation of the training and validation with parallel updates."""  # noqa: D401
+    try:
+      from tqdm.auto import tqdm
+    except ImportError:
+      tqdm = lambda _: _
+    # define dict to contain source and target batch
+    train_batch = {}
+    valid_batch = {}
+
+    # set logging dictionaries
+    train_logs = {"train_loss_f": [], "train_loss_g": [], "train_w_dist": []}
+    valid_logs = {"valid_loss_f": [], "valid_loss_g": [], "valid_w_dist": []}
+
+    for step in tqdm(range(self.num_train_iters)):
+      # execute training steps
+      train_batch["source"] = jnp.array(next(trainloader_source))
+      train_batch["target"] = jnp.array(next(trainloader_target))
+
+      self.state_f, self.state_g, loss, loss_f, loss_g, w_dist = \
+          self.train_step_parallel(
+              self.state_f, self.state_g, train_batch
+          )
+
+      if callback is not None:
+        callback(step, self.to_dual_potentials())
+
+      if not self.pos_weights:
+        self.state_f = self.state_f.replace(
+            params=self._clip_weights_icnn(self.state_f.params)
+        )
+
+      # log to wandb
```

Comment:
is this wandb specific?

Reply:
I didn't add this part, but it doesn't seem like it. Removed in https://github.com/ott-jax/ott/pull/219/commits/827abe4747d74e2580c89342c20374abbd10c0be